### PR TITLE
cli: allow calling core functions directly

### PR DESCRIPTION
**Pull request by helderco**
Fixes https://github.com/dagger/dagger/issues/6947

Current implementation uses a top-level command for core functions:
- `dagger core <function>`  → execute a function from the core API
- `dagger call -m <module> <function>` → execute a function from a module

## How

Can be run from anywhere as it doesn't require a module to load:

```console
❯ dagger core version
[dev-fdc07a5bfb46ca6a37ae379d767210d54f141c4f](version: v0.12.5-010101000000-dev-8e3b1f29b72b)
```

Allows experimentation in the CLI without the need for a module

```console
❯ dagger core container from --address=alpine file --path=/etc/os-release contents
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.20.2
PRETTY_NAME="Alpine Linux v3.20"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://gitlab.alpinelinux.org/alpine/aports/-/issues"
```

Can be a nice alternative to `docker run`:

```console
❯ dagger core container from --address=cgr.dev/chainguard/wolfi-base terminal
dagger / $ 
```

Without arguments:

```console
❯  dagger core
_type: Client
defaultPlatform: linux/arm64
version: v0.12.5-010101000000-dev-8e3b1f29b72b
```

Listing:

```console
❯  dagger core --help
Call a core function

USAGE
  dagger core [options] <function>

FUNCTIONS
  blob                Retrieves a content-addressed blob.
  container           Creates a scratch container.
  current-type-defs   The TypeDef representations of the objects currently being served in the session.
  dagger-engine       The Dagger engine container configuration and state
  default-platform    The default platform of the engine.
  directory           Creates an empty directory.
  git                 Queries a Git repository.
  host                Queries the host environment.
  http                Returns a file containing an http remote url content.
  module              Create a new module.
  module-dependency   Create a new module dependency configuration from a module source and name
  module-source       Create a new module source instance from a source ref string.
  version             Get the current Dagger Engine version.

OPTIONS
  -j, --json            Present result as JSON
  -o, --output string   Save the result to a local file or directory

INHERITED OPTIONS
  -d, --debug             Show debug logs and full verbosity
  -i, --interactive       Spawn a terminal on container exec failure
      --progress string   Progress output format (auto, plain, tty) (default "auto")
  -q, --quiet count       Reduce verbosity (show progress, but clean up at the end)
  -s, --silent            Do not show progress at all
  -v, --verbose count     Increase verbosity (use -vv or -vvv for more)
  -w, --web               Open trace URL in a web browser

Use "dagger core [command] --help" for more information about a command.
```

## Next

The `core` top-level command is possibly a stepping stone to select a few core functions as their own top-level commands. That'll open up creating more commands by implementing them via the API rather than client side. \cc @sipsma

### CODE REVIEW

**Comment by helderco** [comment]
Update: changed implementation from `-m` flag to new `-c,--core` flag. PR this was based on was merged so taking out of draft.

**Comment by helderco** [comment]
Reimplemented as `dagger core`.

File: cmd/dagger/module.go
Comment (helderco) [inline cmd/dagger/module.go:1128]: These are the hardcoded exclusions when you do `dagger core --help`. May want to add more to the list. To be clear, this means these commands won't be created. It's not just for listing in `--help`.

Example of the output with current list in this PR's description.

Comment (helderco) [reply cmd/dagger/module.go:1128]: I think these should be excluded as well:
- set-secret
- cache-volume

I can see a use case for the remaining ones even if just to debug, introspect, or experiment/prototype.

**Review by shykes — APPROVED** [review]
We've been iterating for a while, I think we should merge and make adjustments as needed.

I also like `call --core`, but after discussing with @helderco, `dagger core` has a few advantages.

File: sdk/go/querybuilder/querybuilder.go
Comment (jedevc) [inline sdk/go/querybuilder/querybuilder.go:1]: Trying to understand the querybuilder changes - is this just a refactor/interface change, or is there a behavior change? If the latter, do we need a new test?

File: docs/static/api/reference/index.html
Comment (jedevc) [inline docs/static/api/reference/index.html:1]: We need https://github.com/dagger/dagger/pull/8118 to avoid these random changes :cry:

File: .changes/unreleased/Added-20240629-000342.yaml
Comment (jedevc) [inline .changes/unreleased/Added-20240629-000342.yaml:2]: :pray: could we get another line here? A small example, or even just a redirection to run `dagger core --help`.

**Review by jedevc — APPROVED** [review]
There's a *lot* of cli changes, so hopefully okay to not review in lots of details, but I like the changes that simplify the `FuncCommand` fields, those were tricky before.

Left some small nits, but LGTM.

File: sdk/go/querybuilder/querybuilder.go
Comment (helderco) [reply sdk/go/querybuilder/querybuilder.go:1]: It's a refactor for what was introduced in the printing object fields PR. Only place where it's used (when dagger call returns an object in the CLI).

File: .changes/unreleased/Added-20240629-000342.yaml
Comment (helderco) [reply .changes/unreleased/Added-20240629-000342.yaml]: Added in https://github.com/dagger/dagger/pull/7310/commits/1d36e77d177a6c7d4614feb490dddd7c2f859ea2.

**Comment by helderco** [comment]
> I also like `call --core`, but after discussing with @helderco, `dagger core` has a few advantages.

We can keep the refactoring that allows the stepping stone to `dagger container`. It's a very small change to turn this back into the `call --core` flag. The inverse wasn't true but that work is done and not lost. It would take around 5mins to switch to the flag.

Basically add the mutually exclusive flag (code already in a previous commit in this PR), and set `fc.DisableModuleLoad = true` dynamically if `--core` is set, instead of statically here:

https://github.com/dagger/dagger/blob/1d36e77d177a6c7d4614feb490dddd7c2f859ea2/cmd/dagger/call.go#L20-L29

Perhaps at least mark `dagger core` as _experimental_ which shows a message in `dagger core --help` (the feature is still available, even though we're not using it in any command atm). That would signal to users that it could later change to something else, whether it's `dagger call -c container` or `dagger container`.

**Comment by jedevc** [comment]
> Perhaps at least mark `dagger core` as _experimental_ which shows a message in `dagger core --help` (the feature is still available, even though we're not using it in any command atm)

I think this is worth doing regardless.

I'd also be happier with a `--core` flag - but at this point, we've been discussing for so longer, I really really would just rather ship something in the next release. As you mention, if it's experimental, we can adjust with feedback later :tada:

I've added this to the next milestone, so we'll definitely get this for the next release - at least then we've got an upper limit on how long we're allowed to keep discussing this for :laughing:

File: sdk/go/querybuilder/querybuilder.go
Comment (jedevc) [reply sdk/go/querybuilder/querybuilder.go:1]: Thanks for the explanation, much appreciated :pray:

**Comment by helderco** [comment]
I added the "experimental" annotation. This way it's justifiable if we decide to change the DX. Just waiting for the tests to pass before merge.

**Pull request merged**
